File: lis.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <algorithm>

constexpr int inf = std::numeric_limits<int>::max();

struct lis {
	// Initialize arrays.
	void begin() {
		L.resize(0);
		old.resize(0);
	}

	// Return the index of the first entry greater than v.
	size_t insertion_point(int v) {
		return std::lower_bound(L.begin(), L.end(), v + 1) - L.begin();
	}

	// Return the index of the last entry in L less than v.
	size_t restore_point(int v) {
		return std::lower_bound(L.begin(), L.end(), v) - L.begin() - 1;
	}

	// Return the length of the longest increasing subsequence.
	size_t size() {
		return L.size();
	}

	// Update L array according to next input element.
	void push(int v) {
		size_t j = insertion_point(v);
		if (j == L.size()) {
			old.push_back(inf);
			L.push_back(v);
		} else {
			old.push_back(L[j]);
			L[j] = v;
		}
		print_L();
	}

	// Backtrack, computing the LIS.
	std::vector<int> end() {
		std::vector<int> result(size());
		if (result.size() == 0) return result;
		size_t n = result.size() - 1;
		while (true) {
			result[n] = L[n];
			size_t j = restore_point(old.back());
			if (old.back() == inf) L.pop_back();
			else L[j] = old.back();
			old.pop_back();
			print_L();
			if (j == n)
				if (n-- == 0)
					break;
		}
		return result;
	}

	// For debugging, print out the L array to standard output.
	void print_L() {
		std::cout << "L:";
		for (int v : L) std::cout << ' ' << v;
		std::cout << std::endl;
	}

	// L[j]: Smallest possible last element
	// of an increasing subsequence of length j.
	std::vector<int> L;
	// old[i]: The element of L that was overwritten
	// when reading element x[i]. Allows reconstruction of
	// past L table using binary search.
	std::vector<int> old;
};

int main() {
	lis l;
	l.begin();
	// Read integers from standard input.
	int v;
	while (std::cin >> v) l.push(v);
	// Compute longest increasing subsequence.
	std::vector<int> r = l.end();
	// Output longest increasing subsequence.
	std::cout << r.size() << ":";
	for (int v : r) std::cout << ' ' << v;
	std::cout << std::endl;
	return 0;
}

File: lis.py
GLOBALS = 'inf'.split()
PATTERNS = [('inf', r'\infty')]


def lower_bound(L, v):
    """
    >>> lower_bound([1, 2, 3, 4], 3)
    2
    >>> lower_bound([0, 0, 0], 0)
    0
    >>> lower_bound([0, 0, 0], 1)
    3
    """
    i = 0
    j = len(L)
    while i < j:
        k = i + (j - i) // 2
        if L[k] < v:
            i = k + 1
        else:
            j = k
    return i


inf = float('inf')


def lis(x):
    """
    >>> lis([9, 44, 32, 12, 7, 42, 34, 92, 35, 37, 41, 8, 20, 27, 83, 64, 61,
    ...      28, 39, 93, 29, 17, 13, 14, 55, 21, 66, 72, 23, 73, 99, 1, 2, 88,
    ...      77, 3, 65, 83, 84, 62, 5, 11, 74, 68, 76, 78, 67, 75, 69, 70, 22])
    7
    8
    20
    27
    28
    29
    55
    66
    72
    73
    74
    76
    78
    """
    L = []
    old = []
    for i in range(len(x)):
        j = lower_bound(L, x[i] + 1)
        if j == len(L):
            old.append(inf)
            L.append(x[i])
        else:
            old.append(L[j])
            L[j] = x[i]
    result = [0] * len(L)
    n = len(L)
    while n > 0:
        result[n - 1] = L[n - 1]
        j = lower_bound(L, old[-1]) - 1
        if old[-1] == inf:
            L[-1:] = []
        else:
            L[j] = old[-1]
        old[-1:] = []
        if j == n - 1:
            n = n - 1
    for i in range(len(result)):
        print(result[i])
